### PR TITLE
Offload value deserialization from DB threadpool and add metrics

### LINKED ISSUE #5909: Move serialization outside the database thread
The data written to the index schema is prepared in the database thread including the DAML LF value translation (createArgument, key, choiceArgument, choiceResult). This means a database thread/connection is blocked from being used by another request.

All value translation should be done outside the database thread.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Performance tests revealed that for services reading DAML-LF values off the participant index ~30% of the time is consistently spent deserializing those values.

This PR

1. adds metrics for deserialization time when fetching events and contracts
2. offloads deserialization from the DB thread pool to free up the resource
3. postpones deserialization of values in flat events _after_ filtering out transient contracts

Fixes #5909

CHANGELOG_BEGIN
[Ledger Integration Kit] Added new metrics to time the translation of serialized DAML-LF values when fetched from the participant index.
- ``daml.index.db.get_active_contracts.deserialization`` 
- ``daml.index.db.get_flat_transactions.deserialization`` 
- ``daml.index.db.get_transaction_trees.deserialization`` 
- ``daml.index.db.lookup_transaction_tree_by_id.deserialization`` 
- ``daml.index.db.lookup_flat_transaction_by_id.deserialization`` 
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/ContractsReader.scala [hunk 1]
```diff
@@ -22,20 +24,41 @@ private[dao] sealed abstract class ContractsReader(
...
+            template = Identifier.assertFromString(templateId),
+            arg = Timed.value(
+              lookupActiveContractDeserializationTimer,
+              deserialize(
+                stream = createArgument,
+                errorContext = s"Failed to deserialize create argument for contract $contractId",
+              ),
+            ),
+            agreementText = ""
+          )
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/ContractsReader.scala]: Addressed by https://github.com/digital-asset/daml/pull/5917/commits/c2ee9e27e1fc976d0af7cb44885493f46602400e

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTable.scala [hunk 2]
```diff
@@ -55,13 +53,42 @@ private[events] trait EventsTable {
...
   object Entry {
 
+    private def deserialize[E](
+        rawEntries: Vector[Entry[Raw[E]]],
+        verbose: Boolean,
+        timer: Timer,
+    ): Vector[Entry[E]] =
+      Timed.value(
+        timer,
+        rawEntries.map(entry => entry.copy(event = entry.event.applyDeserialization(verbose))))
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTable.scala]: Addressed by https://github.com/digital-asset/daml/pull/5917/commits/d60c0309b379ff33145a608db55129beb8db7fbf

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTable.scala [hunk 3]
```diff
@@ -55,13 +53,42 @@ private[events] trait EventsTable {
...
+    private def flatTransaction(
+        events: Vector[Entry[Raw.FlatEvent]],
+        verbose: Boolean,
+        deserializationTimer: Timer,
+    ): Option[ApiTransaction] =
       events.headOption.flatMap { first =>
-        val flatEvents =
-          TransactionConversion.removeTransient(events.iterator.map(_.event).toVector)
+        val withoutTransients = removeTransient(events)
+        val flatEvents = deserialize(withoutTransients, verbose, deserializationTimer).map(_.event)
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTable.scala]: Addressed by https://github.com/digital-asset/daml/pull/5917/commits/86c7d7dfba472df0384b8c3d90fb7e793832d8f9

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/Raw.scala [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        raw.copy(
+          contractKey = createKeyValue.map(
+            key =>
+              LfEngineToApi.assertOrRuntimeEx(
+                failureContext = s"attempting to deserialize persisted key to value",
+                LfEngineToApi
+                  .lfVersionedValueToApiValue(
+                    verbose = verbose,
+                    value = ValueSerializer.deserializeValue(key),
+                  ),
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/Raw.scala]: Addressed by https://github.com/digital-asset/daml/pull/5917/commits/953ecfd11f5753e1d853d7ae618ce69983502ef9

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/Raw.scala [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+  Event => PbFlatEvent,
+  ExercisedEvent => PbExercisedEvent
+}
+import com.daml.ledger.api.v1.transaction.{TreeEvent => PbTreeEvent}
+import com.daml.platform.participant.util.LfEngineToApi
+import com.daml.platform.store.serialization.ValueSerializer
+
+sealed trait Raw[+E] {
+  def applyDeserialization(verbose: Boolean): E
+}
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/Raw.scala]: Addressed by https://github.com/digital-asset/daml/pull/5917/commits/b924b0df2cdd7cf19f2aa64e3afdcc719f4fc149

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala [hunk 6]
```diff
@@ -29,8 +30,28 @@ private[dao] final class TransactionsReader(
...
+  private val getFlatTransactionsDeserializationTimer =
+    metrics.daml.index.db.deserialization(getFlatTransactions)
+  private val lookupFlatTransactionByIdDeserializationTimer =
+    metrics.daml.index.db.deserialization(lookupFlatTransactionById)
+  private val getTransactionTreesDeserializationTimer =
+    metrics.daml.index.db.deserialization(getTransactionTrees)
+  private val lookupTransactionTreeByIdDeserializationTimer =
+    metrics.daml.index.db.deserialization(lookupTransactionTreeById)
+  private val getActiveContractsDeserializationTimer =
+    metrics.daml.index.db.deserialization(getActiveContracts)
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala:53]: I think these could potentially be moved to the `Metrics` class.

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala:53]: Good point. Ideally I would also want to pass in the timer as a parameter to `DbDispatcher#runSQL`. Do you think I can tackle it in one go as a separate contribution?

Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala:53]: Fine with me.

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala:53]: Addressed in https://github.com/digital-asset/daml/pull/5927

**Pull request merged**